projects/python-client/src/datapane/client/api/ipython_utils.py

```diff
@@ -19,10 +25,219 @@ def block_to_iframe(block: BaseElement) -> str:
+        Datapane Block
+    """
+    from .report.blocks import Code
+
+    block = Code("".join(cell["source"]))
+
+    return block
+
+
+def output_cell_to_block(cell: dict, ipython_output_cache: dict) -> typing.Any[BaseElement, None]:
+    """Convert a IPython notebook output cell to a Datapane Block
+
+    Args:
+        cell: IPython notebook cell dict
+
+    Returns:
+        Datapane Block
+    """
+    from .report.blocks import wrap_block
+
+    # Get the output object from the IPython output cache
+    cell_output_object = ipython_output_cache.get(cell["execution_count"], None)
+
+    # If there's corresponding output object, skip
+    if cell_output_object is None:
+        return None
+
+    try:
+        block = wrap_block(cell_output_object)
+        return block
+    except Exception:
+        return None
+
+
+@capture_event("IPython Cells to Blocks")
+def cells_to_blocks(ipython_output_cache: dict, opt_out: bool = True) -> typing.List[BaseElement]:
+    """Convert IPython notebook cells to a list of Datapane Blocks
+
+    Recognized cell tags:
+        - `dp-exclude` - Exclude this cell (when opt_out=True)
+        - `dp-include` - Include this cell (when opt_out=False)
+        - `dp-show-code` - Show the input code for this cell
+
+    Args:
+        ipython_output_cache: The output cache (Out or _oh) dict from a IPython notebook
+        opt_out: When True, all cells are converted to blocks unless explicitly opted out with the tag `dp-exclude`. When False, only cells with the tag `dp-include` are converted to blocks.
+
+    Returns:
+        List of Datapane Blocks
+
+    ..note:: IPython output caching must be enabled for this function to work. It is enabled by default.
+    """
+    display_msg("Converting cells to blocks.")
+
+    notebook_json = get_notebook_json()
+
+    blocks = []
+
+    for cell in notebook_json["cells"]:
+        tags = cell["metadata"].get("tags", [])
+
+        if (opt_out and "dp-exclude" not in tags) or (not opt_out and "dp-include" in tags):
+            block = None  # type: typing.Any[BaseElement, None]
+
+            if cell["cell_type"] == "markdown":
+                block = markdown_cell_to_block(cell)
+                blocks.append(block)
+
+            elif cell["cell_type"] == "code":
+                if "dp-show-code" in tags:
+                    block = input_cell_to_block(cell)
+                    blocks.append(block)
+
+                if cells_to_blocks.__name__ not in "".join(cell["source"]):
+                    block = output_cell_to_block(cell, ipython_output_cache)
+                    if block:
+                        blocks.append(block)
+                    elif "dp-include" in tags:
+                        display_msg(
+                            f'Cell output of type {type(ipython_output_cache.get(cell["execution_count"]))} not supported. Skipping.',
+                        )
+
+    if not blocks:
+        display_msg("No blocks found.")
+
+    display_msg(
```

Comment:
It might be a bit clearer if this message is displayed at the start of the function

Reply:
That was the original order! I changed it as it made more sense to me as the final message when there were errors. I've put it back at the start now!